community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java

```diff
@@ -611,34 +616,63 @@ private InternalTransaction beginTransactionInternal( KernelTransaction.Type typ
 
     private ResourceIterator<Node> nodesByLabelAndProperty( Label myLabel, String key, Value value )
     {
-        Statement statement = statementContext.get();
-
-        ReadOperations readOps = statement.readOperations();
-        int propertyId = readOps.propertyKeyGetForName( key );
-        int labelId = readOps.labelGetForName( myLabel.name() );
+        KernelTransaction transaction = statementContext.getKernelTransactionBoundToThisThread( true );
+        Statement statement = transaction.acquireStatement();
+        Read read = transaction.dataRead();
+        TokenRead tokenRead = transaction.tokenRead();
+        int propertyId = tokenRead.propertyKey( key );
+        int labelId = tokenRead.labelGetForName( myLabel.name() );
 
         if ( propertyId == NO_SUCH_PROPERTY_KEY || labelId == NO_SUCH_LABEL )
         {
             statement.close();
             return emptyResourceIterator();
         }
+        CapableIndexReference index = transaction.schemaRead().index( labelId, propertyId );
+        if ( index != CapableIndexReference.NO_INDEX )
+        {
+            // Ha! We found an index - let's use it to find matching nodes
+            try
+            {
+                NodeValueIndexCursor cursor = transaction.cursors().allocateNodeValueIndexCursor();
+                IndexQuery.ExactPredicate query = IndexQuery.exact( propertyId, value );
+                read.nodeIndexSeek( index, cursor, IndexOrder.NONE, query );
 
-        IndexDescriptor descriptor = findAnyIndexByLabelAndProperty( readOps, propertyId, labelId );
+                return new PrefetchingResourceIterator<Node>()
+                {
+                    private boolean closed;
+                    @Override
+                    protected Node fetchNextOrNull()
+                    {
+                        if ( cursor.next() )
+                        {
+                            return newNodeProxy( cursor.nodeReference() );
+                        }
+                        else
+                        {
+                            close();
+                            return null;
+                        }
+                    }
 
-        try
-        {
-            if ( null != descriptor )
+                    @Override
+                    public void close()
+                    {
+                        if ( !closed )
+                        {
+                            statement.close();
+                            cursor.close();
```

Comment:
Use `IOUtil.closeAll` (or maybe introduce a `closeBoth` if you don't want to allocate the variadic array) or something like that, to ensure that the cursor is closed, even if `statement.close` throws an exception.